doc/tut3.rst

```diff
@@ -0,0 +1,277 @@
+``int``.
+
+Code blocks as arguments
+------------------------
+
+In Nim it is possible to pass the last argument of a call expression in a
+separate code block with indentation. For example the following code
+example is a valid (but not a recommended) way to call ``echo``:
+
+.. code-block:: nim
+
+  echo "Hello ":
+    let a = "Wor"
+    let b = "ld!"
+    a & b
+
+For macros this way of calling is useful for example to implement an
+embedded domain specific language. Syntax trees of arbitrary
+complexity can be passed to macros with this notation.
+
+
+The Syntax Tree
+---------------
+
+In order to build a Nim syntax tree one needs to know how Nim source
+code is represented as a syntax tree, and how such a tree needs to
+look like so that the Nim compiler will understand it. The nodes of the
+Nim syntax tree are documented in the `macros <macros.html>`_ module.
+But a probably more interesting and interactive way to explore the Nim
+syntax tree is with ``macros.treeRepr``, it converts a syntax tree
+into a multi line string for printing on the console.  It can be used
+to explore how the argument expressions are represented in tree form
+and for debug printing of generated syntax tree.
+
+
+Custom sematic checking
+-----------------------
+
+The first thing that a macro should do with it's arguments is to check
+if the argument is in the correct form.  Not every type of wrong input
+needs to be caught here, but anything that could cause a crash during
+macro evaluation should be caught and create a nice error message.
+``macros.expectKind`` and ``macros.expectLen`` are a good start.  If
+the checks need to be more complex, arbitrary error messages can
+be created with the ``macros.error`` proc.
+
+Generating Code
+---------------
+
+There are two ways to generate the code. Either by creating the syntax
+tree with expressions that contain a lot of calls to ``newTree`` and
+``newLit``, or with ``quote do:`` expressions. This offers the best low
+level control for the syntax tree generation, but the second option is
+much less verbose.  If you choose to create the syntax tree with calls
+to ``newTree`` and ``newLit`` the macro ``marcos.dumpAstGen`` can help
+you with the verbosity. ``quote do:`` allows you to write the code
+that you want to generate literally, backticks are used to insert code
+from ``NimNode`` symbols into the generated expression. This means
+that you can't use backticks within ``quote do:`` for anything else.
+
+
+Building your first macro
+-------------------------
+
+To give a footstart to writing macros we will show now how to
+implement the ``myDebug`` macro mentioned earlier.  The first thing to
+do is to build a simple example of the macro usage, and then just
+print the argument.  This way it is possible to get an idea of a
+correct argument should be look like.
+
+.. code-block:: nim
+    :test: "nim c $1"
+
+  import macros
+
+  macro myAssert(arg: untyped): untyped =
+    echo arg.treeRepr
```

Comment:
Why do we use `echo arg.treeRepr` inside of a macro, what are the advantages over:

```nim
dumpTree:
  a != b
```